src/email/outbound.js

```diff
@@ -0,0 +1,303 @@
+const { createConnection } = require('net');
+const { resolveMx } = require('dns');
+
+const { DKIMSign } = require('dkim-signer');
+const MailComposer = require("nodemailer/lib/mail-composer");
+const ParseEmailAddress = require("email-addresses");
+const { Logging } = require('../log');
+
+const CRLF = '\r\n';
+
+const log = Logging.get("outbound");
+
+module.exports = function (options) {
+    options = options || {};
+
+    const dkimPrivateKey = (options.dkim || {}).privateKey;
+    const dkimKeySelector = (options.dkim || {}).keySelector || 'dkim';
+    const devPort = options.devPort || -1;
+    const devHost = options.devHost || 'localhost';
+    const smtpPort = options.smtpPort || 25;
+    const smtpHost = options.smtpHost || -1;
+
+
+    function groupRecipients(recipients) {
+        let groups = {};
+        let host;
+        const recipientsLength = recipients.length;
+        for (let i = 0; i < recipientsLength; i++) {
+            host = ParseEmailAddress.parseOneAddress(recipients[i]).domain;
+            (groups[host] || (groups[host] = [])).push(recipients[i]);
+        }
+        return groups;
+    }
+
+    /**
+     * connect to domain by MX record
+     */
+    function connectMx(domain, callback) {
```

Comment:
I would consider trying to make this a promise based function using https://nodejs.org/api/dns.html#dns_dns_promises_api. If you used the async/await syntax, you could reduce the indentation of this function by a lot, which makes for an easier read.